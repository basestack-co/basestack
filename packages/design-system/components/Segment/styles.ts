import styled from "styled-components";
import { space } from "styled-system";

export const Container = styled.div.withConfig({
  shouldForwardProp: (prop) => prop !== "backgroundColor",
})<{
  backgroundColor?: string;
}>`
  ${space};
  display: flex;
  background-color: ${({ theme, backgroundColor }) =>
    backgroundColor || theme.segment.backgroundColor};
  border-radius: 4px;
  height: 36px;
  padding: 4px;
  overflow: hidden;
`;

export const Wrapper = styled.div`
  position: relative;
  display: flex;
  align-items: center;
`;

export const Button = styled.button`
  border: none;
  background-color: transparent;
  cursor: pointer;
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 12px;
  z-index: 1;
  border-radius: 4px;
  transition: background-color 0.1s ease-in-out;

  &:hover:not(.active) {
    background-color: ${({ theme }) =>
      theme.segment.button.hover.backgroundColor};
  }
`;

export const Slider = styled.div.withConfig({
  shouldForwardProp: (prop) => prop !== "translateX",
})<{
  translateX: number | string;
  $width: number;
}>`
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  pointer-events: none;
  z-index: 0;

  &::before {
    position: absolute;
    content: "";
    top: 0;
    bottom: 0;
    background-color: ${({ theme }) => theme.segment.slider.backgroundColor};
    transition:
      left 0.2s ease-in-out,
      width 0.2s ease-in-out;
    width: ${({ $width }) => `${$width}px`};
    border-radius: 4px;
    box-shadow: ${({ theme }) => theme.shadow.elevation2};
    left: ${({ translateX }) => `${translateX}px`};
  }
`;
