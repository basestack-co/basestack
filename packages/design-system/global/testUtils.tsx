import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
