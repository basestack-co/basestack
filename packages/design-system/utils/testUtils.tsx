import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
