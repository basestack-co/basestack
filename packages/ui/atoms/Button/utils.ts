import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
