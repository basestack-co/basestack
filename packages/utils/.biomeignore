# Node dependencies
node_modules

# Next.js build output
.next
out

# Environment files
.env
.env.*
!.env.example

# Logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
pnpm-debug.log*

# OS / editor files
.DS_Store
Thumbs.db

# Package manager cache
.pnp
.pnp.js
.pnp.cjs

# Test coverage
coverage

# Build artifacts
dist
build

# Temporary files
.tmp
temp