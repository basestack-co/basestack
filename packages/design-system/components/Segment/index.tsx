import { useReducer, useRef, useEffect, memo } from "react";
import { SpaceProps } from "styled-system";
import { useTheme } from "styled-components";
import Text from "../Text";
import Icon from "../Icon";
import { Container, Button, Slider, Wrapper } from "./styles";

type Item = {
  label?: string;
  text?: string;
  icon?: string;
  id: string;
};

export interface SegmentProps extends SpaceProps {
  /**
   * List of items
   */
  items: Array<Item>;
  /**
   * Selected item
   */
  onSelect: (selected: string) => void;
  selectedIndex?: number;
}

const reducer = (state: any, action: { type: any; payload: any }) => {
  switch (action.type) {
    case "SET_SELECTED_VALUES":
      return {
        ...state,
        selected: action.payload.selected,
        translateX: action.payload.translateX,
        width: action.payload.width,
      };
    default:
      return state;
  }
};

const Segment = ({
  items,
  onSelect,
  selectedIndex,
  ...props
}: SegmentProps) => {
  const theme = useTheme();
  const buttonRefs = useRef<(HTMLButtonElement | null)[]>([]);

  const [{ selected, translateX, width }, dispatch] = useReducer(reducer, {
    selected: selectedIndex || 0,
    translateX: 0,
    width: 0,
  });

  useEffect(() => {
    if (buttonRefs.current[selected]) {
      const selectedButton = buttonRefs.current[selected];
      const buttonWidth = selectedButton ? selectedButton.offsetWidth : 0;
      const buttonLeft = selectedButton ? selectedButton.offsetLeft : 0;

      dispatch({
        type: "SET_SELECTED_VALUES",
        payload: { selected, translateX: buttonLeft, width: buttonWidth },
      });
    }
  }, [selected, items]);

  return (
    <Container data-testid="segment-component" {...props}>
      <Wrapper>
        {items.map((item: Item, index: number) => {
          const buttonProps = selected === index ? { className: "active" } : {};
          return (
            <Button
              key={index.toString()}
              data-testid="segment-button"
              {...buttonProps}
              // @ts-ignore
              ref={(el) => (buttonRefs.current[index] = el)}
              onClick={() => {
                onSelect(item.id);
                dispatch({
                  type: "SET_SELECTED_VALUES",
                  payload: { selected: index, translateX, width },
                });
              }}
            >
              {!!item.icon && (
                <Icon icon={item.icon} color={theme.segment.icon.color} />
              )}
              {!!item.text && (
                <Text
                  size="small"
                  ml={!!item.icon ? theme.spacing.s1 : 0}
                  flexShrink={0}
                >
                  {item.text}
                </Text>
              )}
              {!!item.label && (
                <Text
                  size="small"
                  ml={theme.spacing.s1}
                  flexShrink={0}
                  color={
                    theme.segment.label[selected === index ? "active" : "color"]
                  }
                >
                  {item.label}
                </Text>
              )}
            </Button>
          );
        })}
        <Slider $width={width} translateX={translateX} />
      </Wrapper>
    </Container>
  );
};

export default memo(Segment);
