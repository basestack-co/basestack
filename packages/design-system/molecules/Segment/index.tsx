import { useReducer, memo } from "react";
import { SpaceProps } from "styled-system";
import { useTheme } from "styled-components";
import { Text, Icon } from "../../atoms";
import { Container, Button, Slider } from "./styles";

type Item = {
  text?: string;
  icon?: string;
  id: string;
};

export interface SegmentProps extends SpaceProps {
  /**
   * List of items
   */
  items: Array<Item>;
  /**
   * Selected item
   */
  onSelect: (selected: string) => void;
  selectedIndex?: number;
}

const reducer = (
  state: any,
  action: { type: any; payload: { selected: number; translateX: number } }
) => {
  switch (action.type) {
    case "SET_SELECTED_VALUES":
      return {
        ...state,
        selected: action.payload.selected,
        translateX: action.payload.translateX,
      };
    default:
      return state;
  }
};

const Segment = ({
  items,
  onSelect,
  selectedIndex,
  ...props
}: SegmentProps) => {
  const theme = useTheme();
  const [{ selected, translateX }, dispatch] = useReducer(reducer, {
    selected: 0,
    translateX: (selectedIndex || 0) * 100,
  });

  return (
    <Container data-testid="segment-component" {...props}>
      {items.map((item: Item, index: number) => {
        const buttonProps = selected === index ? { className: "active" } : {};
        return (
          <Button
            key={index.toString()}
            data-testid="segment-button"
            {...buttonProps}
            onClick={() => {
              dispatch({
                type: "SET_SELECTED_VALUES",
                payload: { selected: index, translateX: index * 100 },
              });
              onSelect(item.id);
            }}
          >
            {!!item.icon && (
              <Icon icon={item.icon} color={theme.colors.black} />
            )}
            {!!item.text && (
              <Text size="small" ml={!!item.icon ? theme.spacing.s1 : 0}>
                {item.text}
              </Text>
            )}
          </Button>
        );
      })}
      <Slider translateX={translateX} numberOfItems={items.length} />
    </Container>
  );
};

export default memo(Segment);
