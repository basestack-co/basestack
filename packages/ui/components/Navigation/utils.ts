// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
