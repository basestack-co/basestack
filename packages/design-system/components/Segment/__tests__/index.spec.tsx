import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Segment from "..";

describe("Segment tests", () => {
  afterEach(cleanup);

  test("should render Segment correctly", () => {
    const { asFragment } = renderWithTheme(
      <Segment
        onSelect={jest.fn()}
        items={[
          { icon: "view_module", id: "cards" },
          { icon: "view_stream", id: "table" },
        ]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render Segment with icons and test", () => {
    const { asFragment } = renderWithTheme(
      <Segment
        onSelect={jest.fn()}
        items={[
          { text: "Item 1", icon: "view_module", id: "cards" },
          { text: "Item 2", icon: "view_stream", id: "table" },
        ]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render active Segment", () => {
    const { container, getAllByTestId } = renderWithTheme(
      <Segment
        onSelect={jest.fn()}
        items={[
          { text: "Item 1", id: "cards" },
          { text: "Item 2", id: "table" },
        ]}
      />,
    );
    const item1 = getAllByTestId("segment-button")[0];

    expect(item1).toHaveClass("active");
  });
});
