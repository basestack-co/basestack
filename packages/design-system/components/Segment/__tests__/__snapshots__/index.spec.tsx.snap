// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segment tests should render Segment correctly 1`] = `
<DocumentFragment>
  .c3 {
  height: 1.5rem;
  width: 1.5rem;
  flex-shrink: 0;
}

.c4 {
  color: #000000;
  font-size: 1.5rem;
}

.c0 {
  display: flex;
  background-color: #EEEEEE;
  border-radius: 4px;
  height: 36px;
  padding: 4px;
  overflow: hidden;
}

.c1 {
  position: relative;
  display: flex;
  align-items: center;
}

.c2 {
  border: none;
  background-color: transparent;
  cursor: pointer;
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 12px;
  z-index: 1;
  border-radius: 4px;
  transition: background-color 0.1s ease-in-out;
}

.c2:hover:not(.active) {
  background-color: #E2E2E2;
}

.c5 {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  pointer-events: none;
  z-index: 0;
}

.c5::before {
  position: absolute;
  content: "";
  top: 0;
  bottom: 0;
  background-color: #FFFFFF;
  transition: left 0.2s ease-in-out,width 0.2s ease-in-out;
  width: 0px;
  border-radius: 4px;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px,rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
  left: 0px;
}

<div
    class="c0"
    data-testid="segment-component"
  >
    <div
      class="c1"
    >
      <button
        class="c2 active"
        data-testid="segment-button"
      >
        <div
          class="c3"
          data-testid="icon-container"
          size="24"
        >
          <span
            class="c4 material-symbols-rounded"
            color="#000000"
            data-testid="icon"
            size="24"
          >
            view_module
          </span>
        </div>
      </button>
      <button
        class="c2"
        data-testid="segment-button"
      >
        <div
          class="c3"
          data-testid="icon-container"
          size="24"
        >
          <span
            class="c4 material-symbols-rounded"
            color="#000000"
            data-testid="icon"
            size="24"
          >
            view_stream
          </span>
        </div>
      </button>
      <div
        class="c5"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Segment tests should render Segment with icons and test 1`] = `
<DocumentFragment>
  .c5 {
  font-size: 0.875rem;
  line-height: 1.375rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  word-break: break-word;
  color: #000000;
  font-family: roboto;
  margin-left: 0.25rem;
  flex-shrink: 0;
}

.c3 {
  height: 1.5rem;
  width: 1.5rem;
  flex-shrink: 0;
}

.c4 {
  color: #000000;
  font-size: 1.5rem;
}

.c0 {
  display: flex;
  background-color: #EEEEEE;
  border-radius: 4px;
  height: 36px;
  padding: 4px;
  overflow: hidden;
}

.c1 {
  position: relative;
  display: flex;
  align-items: center;
}

.c2 {
  border: none;
  background-color: transparent;
  cursor: pointer;
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 12px;
  z-index: 1;
  border-radius: 4px;
  transition: background-color 0.1s ease-in-out;
}

.c2:hover:not(.active) {
  background-color: #E2E2E2;
}

.c6 {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  pointer-events: none;
  z-index: 0;
}

.c6::before {
  position: absolute;
  content: "";
  top: 0;
  bottom: 0;
  background-color: #FFFFFF;
  transition: left 0.2s ease-in-out,width 0.2s ease-in-out;
  width: 0px;
  border-radius: 4px;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px,rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
  left: 0px;
}

<div
    class="c0"
    data-testid="segment-component"
  >
    <div
      class="c1"
    >
      <button
        class="c2 active"
        data-testid="segment-button"
      >
        <div
          class="c3"
          data-testid="icon-container"
          size="24"
        >
          <span
            class="c4 material-symbols-rounded"
            color="#000000"
            data-testid="icon"
            size="24"
          >
            view_module
          </span>
        </div>
        <p
          class="c5"
          font-family="roboto"
          ml="0.25rem"
        >
          Item 1
        </p>
      </button>
      <button
        class="c2"
        data-testid="segment-button"
      >
        <div
          class="c3"
          data-testid="icon-container"
          size="24"
        >
          <span
            class="c4 material-symbols-rounded"
            color="#000000"
            data-testid="icon"
            size="24"
          >
            view_stream
          </span>
        </div>
        <p
          class="c5"
          font-family="roboto"
          ml="0.25rem"
        >
          Item 2
        </p>
      </button>
      <div
        class="c6"
      />
    </div>
  </div>
</DocumentFragment>
`;
